File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Observer.WeatherStation/WeatherStationHeatIndex.cs
using System;
using System.Collections.Generic;

namespace DoFactory.HeadFirst.Observer.WeatherStation
{
    class WeatherStationHeatIndex
    {
        static void Main(string[] args)
        {
            WeatherData weatherData = new WeatherData();

            CurrentConditionsDisplay currentDisplay = new CurrentConditionsDisplay(weatherData);
            StatisticsDisplay statisticsDisplay = new StatisticsDisplay(weatherData);
            ForecastDisplay forecastDisplay = new ForecastDisplay(weatherData);
            HeatIndexDisplay heatIndexDisplay = new HeatIndexDisplay(weatherData);

            weatherData.SetMeasurements(80, 65, 30.4f);
            weatherData.SetMeasurements(82, 70, 29.2f);
            weatherData.SetMeasurements(78, 90, 29.2f);

            // Wait for user
            Console.ReadKey();
        }
    }

    #region Subject

    public interface ISubject
    {
        void RegisterObserver(IObserver observer);
        void RemoveObserver(IObserver observer);
        void NotifyObservers();
    }

    public class WeatherData : ISubject
    {
        private List<IObserver> _observers = new List<IObserver>();
        private float _temperature;
        private float _humidity;
        private float _pressure;

        public void RegisterObserver(IObserver observer)
        {
            _observers.Add(observer);
        }

        public void RemoveObserver(IObserver observer)
        {
            _observers.Remove(observer);
        }

        public void NotifyObservers()
        {
            foreach (IObserver observer in _observers)
            {
                observer.Update(_temperature, _humidity, _pressure);
            }
        }

        public void MeasurementsChanged()
        {
            NotifyObservers();
        }

        public void SetMeasurements(float temperature, float humidity, float pressure)
        {
            this._temperature = temperature;
            this._humidity = humidity;
            this._pressure = pressure;
            MeasurementsChanged();
        }
    }
    #endregion

    #region Observer

    public interface IObserver
    {
        void Update(float temperature, float humidity, float pressure);
    }

    public interface IDisplayElement
    {
        void Display();
    }

    public class CurrentConditionsDisplay : IObserver, IDisplayElement
    {
        private float _temperature;
        private float _humidity;
        private ISubject _weatherData;

        public CurrentConditionsDisplay(ISubject weatherData)
        {
            this._weatherData = weatherData;
            weatherData.RegisterObserver(this);
        }

        public void Update(float temperature, float humidity, float pressure)
        {
            this._temperature = temperature;
            this._humidity = humidity;
            Display();
        }

        public void Display()
        {
            Console.WriteLine("Current conditions: " + _temperature
                + "F degrees and " + _humidity + "% humidity");
        }
    }

    public class ForecastDisplay : IObserver, IDisplayElement
    {
        private float _currentPressure = 29.92f;
        private float _lastPressure;
        private WeatherData _weatherData;

        public ForecastDisplay(WeatherData weatherData)
        {
            this._weatherData = weatherData;
            weatherData.RegisterObserver(this);
        }

        public void Update(float temperature, float humidity, float pressure)
        {
            _lastPressure = _currentPressure;
            _currentPressure = pressure;

            Display();
        }

        public void Display()
        {
            Console.Write("Forecast: ");

            if (_currentPressure > _lastPressure)
            {
                Console.WriteLine("Improving weather on the way!");
            }
            else if (_currentPressure == _lastPressure)
            {
                Console.WriteLine("More of the same");
            }
            else if (_currentPressure < _lastPressure)
            {
                Console.WriteLine("Watch out for cooler, rainy weather");
            }
        }
    }

    public class HeatIndexDisplay : IObserver, IDisplayElement
    {
        private float _heatIndex = 0.0f;
        private WeatherData _weatherData;

        public HeatIndexDisplay(WeatherData weatherData)
        {
            this._weatherData = weatherData;
            weatherData.RegisterObserver(this);
        }

        public void Update(float temperature, float humidity, float pressure)
        {
            _heatIndex = ComputeHeatIndex(temperature, humidity);
            Display();
        }

        private float ComputeHeatIndex(float t, float rh)
        {
            float heatindex = (float)
                    (
                    (16.923 + (0.185212 * t)) +
                    (5.37941 * rh) -
                    (0.100254 * t * rh) +
                    (0.00941695 * (t * t)) +
                    (0.00728898 * (rh * rh)) +
                    (0.000345372 * (t * t * rh)) -
                    (0.000814971 * (t * rh * rh)) +
                    (0.0000102102 * (t * t * rh * rh)) -
                    (0.000038646 * (t * t * t)) +
                    (0.0000291583 * (rh * rh * rh)) +
                    (0.00000142721 * (t * t * t * rh)) +
                    (0.000000197483 * (t * rh * rh * rh)) -
                    (0.0000000218429 * (t * t * t * rh * rh)) +
                    (0.000000000843296 * (t * t * rh * rh * rh)) -
                    (0.0000000000481975 * (t * t * t * rh * rh * rh)));
            return heatindex;
        }

        public void Display()
        {
            Console.WriteLine("Heat index is " + _heatIndex + "\n");
        }
    }

    public class StatisticsDisplay : IObserver, IDisplayElement
    {
        private float _maxTemp = 0.0f;
        private float _minTemp = 200;
        private float _tempSum = 0.0f;
        private int _numReadings;
        private WeatherData _weatherData;

        public StatisticsDisplay(WeatherData weatherData)
        {
            this._weatherData = weatherData;
            weatherData.RegisterObserver(this);
        }

        public void Update(float temperature, float humidity, float pressure)
        {
            _tempSum += temperature;
            _numReadings++;

            if (temperature > _maxTemp)
            {
                _maxTemp = temperature;
            }

            if (temperature < _minTemp)
            {
                _minTemp = temperature;
            }

            Display();
        }

        public void Display()
        {
            Console.WriteLine("Avg/Max/Min temperature = " + (_tempSum / _numReadings)
                + "/" + _maxTemp + "/" + _minTemp);
        }
    }

    #endregion
}

File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Observer.WeatherStation/WeatherStationHeatIndex.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic

Namespace DoFactory.HeadFirst.Observer.WeatherStation
	Friend Class WeatherStationHeatIndex
        Shared Sub Main(ByVal args() As String)

            Dim weatherData As New WeatherData()

            Dim currentDisplay As New CurrentConditionsDisplay(weatherData)
            Dim statisticsDisplay As New StatisticsDisplay(weatherData)
            Dim forecastDisplay As New ForecastDisplay(weatherData)
            Dim heatIndexDisplay As New HeatIndexDisplay(weatherData)

            weatherData.SetMeasurements(80, 65, 30.4F)
            weatherData.SetMeasurements(82, 70, 29.2F)
            weatherData.SetMeasurements(78, 90, 29.2F)

            ' Wait for user
            Console.ReadKey()
        End Sub
	End Class

	#Region "Subject"

	Public Interface ISubject
		Sub RegisterObserver(ByVal observer As IObserver)
		Sub RemoveObserver(ByVal observer As IObserver)
		Sub NotifyObservers()
	End Interface

	Public Class WeatherData
		Implements ISubject
		Private _observers As List(Of IObserver) = New List(Of IObserver)()
		Private _temperature As Single
		Private _humidity As Single
		Private _pressure As Single

		Public Sub RegisterObserver(ByVal observer As IObserver) Implements ISubject.RegisterObserver
			_observers.Add(observer)
		End Sub

		Public Sub RemoveObserver(ByVal observer As IObserver) Implements ISubject.RemoveObserver
			_observers.Remove(observer)
		End Sub

		Public Sub NotifyObservers() Implements ISubject.NotifyObservers
			For Each observer As IObserver In _observers
				observer.Update(_temperature, _humidity, _pressure)
			Next observer
		End Sub

		Public Sub MeasurementsChanged()
			NotifyObservers()
		End Sub

		Public Sub SetMeasurements(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single)
			Me._temperature = temperature
			Me._humidity = humidity
			Me._pressure = pressure
			MeasurementsChanged()
		End Sub
	End Class
	#End Region

	#Region "Observer"

	Public Interface IObserver
		Sub Update(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single)
	End Interface

	Public Interface IDisplayElement
		Sub Display()
	End Interface

	Public Class CurrentConditionsDisplay
		Implements IObserver, IDisplayElement
		Private _temperature As Single
		Private _humidity As Single
		Private _weatherData As ISubject

		Public Sub New(ByVal weatherData As ISubject)
			Me._weatherData = weatherData
			weatherData.RegisterObserver(Me)
		End Sub

		Public Sub Update(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single) Implements IObserver.Update
			Me._temperature = temperature
			Me._humidity = humidity
			Display()
		End Sub

		Public Sub Display() Implements IDisplayElement.Display
			Console.WriteLine("Current conditions: " & _temperature & "F degrees and " & _humidity & "% humidity")
		End Sub
	End Class

	Public Class ForecastDisplay
		Implements IObserver, IDisplayElement
		Private _currentPressure As Single = 29.92f
		Private _lastPressure As Single
		Private _weatherData As WeatherData

		Public Sub New(ByVal weatherData As WeatherData)
			Me._weatherData = weatherData
			weatherData.RegisterObserver(Me)
		End Sub

		Public Sub Update(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single) Implements IObserver.Update
			_lastPressure = _currentPressure
			_currentPressure = pressure

			Display()
		End Sub

		Public Sub Display() Implements IDisplayElement.Display
			Console.Write("Forecast: ")

			If _currentPressure > _lastPressure Then
				Console.WriteLine("Improving weather on the way!")
			ElseIf _currentPressure = _lastPressure Then
				Console.WriteLine("More of the same")
			ElseIf _currentPressure < _lastPressure Then
				Console.WriteLine("Watch out for cooler, rainy weather")
			End If
		End Sub
	End Class

	Public Class HeatIndexDisplay
		Implements IObserver, IDisplayElement
		Private _heatIndex As Single = 0.0f
		Private _weatherData As WeatherData

		Public Sub New(ByVal weatherData As WeatherData)
			Me._weatherData = weatherData
			weatherData.RegisterObserver(Me)
		End Sub

		Public Sub Update(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single) Implements IObserver.Update
			_heatIndex = ComputeHeatIndex(temperature, humidity)
			Display()
		End Sub

		Private Function ComputeHeatIndex(ByVal t As Single, ByVal rh As Single) As Single
			Dim heatindex As Single = CSng((16.923 + (0.185212 * t)) + (5.37941 * rh) - (0.100254 * t * rh) + (0.00941695 * (t * t)) + (0.00728898 * (rh * rh)) + (0.000345372 * (t * t * rh)) - (0.000814971 * (t * rh * rh)) + (0.0000102102 * (t * t * rh * rh)) - (0.000038646 * (t * t * t)) + (0.0000291583 * (rh * rh * rh)) + (0.00000142721 * (t * t * t * rh)) + (0.000000197483 * (t * rh * rh * rh)) - (0.0000000218429 * (t * t * t * rh * rh)) + (0.000000000843296 * (t * t * rh * rh * rh)) - (0.0000000000481975 * (t * t * t * rh * rh * rh)))
			Return heatindex
		End Function

		Public Sub Display() Implements IDisplayElement.Display
            Console.WriteLine("Heat index is " & _heatIndex & Constants.vbLf)
		End Sub
	End Class

	Public Class StatisticsDisplay
		Implements IObserver, IDisplayElement
		Private _maxTemp As Single = 0.0f
		Private _minTemp As Single = 200
		Private _tempSum As Single = 0.0f
		Private _numReadings As Integer
		Private _weatherData As WeatherData

		Public Sub New(ByVal weatherData As WeatherData)
			Me._weatherData = weatherData
			weatherData.RegisterObserver(Me)
		End Sub

		Public Sub Update(ByVal temperature As Single, ByVal humidity As Single, ByVal pressure As Single) Implements IObserver.Update
			_tempSum += temperature
			_numReadings += 1

			If temperature > _maxTemp Then
				_maxTemp = temperature
			End If

			If temperature < _minTemp Then
				_minTemp = temperature
			End If

			Display()
		End Sub

		Public Sub Display() Implements IDisplayElement.Display
			Console.WriteLine("Avg/Max/Min temperature = " & (_tempSum / _numReadings) & "/" & _maxTemp & "/" & _minTemp)
		End Sub
	End Class

	#End Region
End Namespace